File: Algorithm/双指针/左右指针/125.验证回文串/125.验证回文串.py
# 125 验证回文串

# 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写
# 说明：本题中，我们将空字符串定义为有效的回文串

import re


class Solution:

    def isPalindrome(self, s: str) -> bool:
        pattern = re.compile(r'[A-Za-z0-9]+')
        arr = pattern.findall(s.upper())
        str = ''.join(arr)
        print(str)
        left = 0
        right = len(str) - 1
        while left <= right:
            if str[left] == str[right]:
                left = left + 1
                right = right - 1
            else:
                return False
        return True

# test
str = 'A man, a plan, a canal: Panama'
solution = Solution()
res = solution.isPalindrome(str)
print(res)

File: Data Structure & Algorithm/Archive/Algorithm/Two Points/zh/左右指针/125.验证回文串/125.验证回文串.ts
// 125 验证回文串

/**
 * 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写
 * 说明：本题中，我们将空字符串定义为有效的回文串
 */

export function isPalindrome(s: string): boolean {
  if (s.length === 0 || s.trim() === '') {
    return true;
  }
  let str: string | undefined = s
    .match(/[A-Za-z0-9]+/g)
    ?.join('')
    .toLowerCase();
  if (str === undefined) {
    return true;
  }
  let left: number = 0;
  let right: number = str.length - 1;
  while (left < right) {
    if (str[left] !== str[right]) {
      return false;
    }
    left++;
    right--;
  }
  return true;
}

// test
const res = isPalindrome('A man, a plan, a canal: Panama');
const res2 = isPalindrome('race a car');
const res3 = isPalindrome('.');
console.log(res);
console.log(res2);
console.log(res3);
